### PR TITLE
feat: client abstracts hostclient layer

### PR DESCRIPTION
**Pull request by Duslia**
#### What type of PR is this?

feat

#### What this PR does / why we need it (English/Chinese):

en: client abstracts protocol layer for http2 and ALPN. Client removes ReadBufferSize and WriteBufferSize config.
zh: client 为 ALPN 和 http2 抽象出协议层 HostClient。client 删除 readbuffersize 和 writebuffersize 配置项。

### CODE REVIEW

File: pkg/app/client/client.go [hunk 1]
```diff
@@ -225,153 +226,50 @@ var defaultClient = Client{DialTimeout: consts.DefaultDialTimeout}
...
 // Don't SetProxy twice for a client.
 // If you want to use another proxy, please create another client and set proxy to it.
 func (c *Client) SetProxy(p protocol.Proxy) {
-	c.Proxy = p
+	c.proxy = p
+}
+
+// SetRetryIf is used to set client proxy.
+//
+// Don't SetProxy twice for a client.
```
Comment (welkeyever) [inline pkg/app/client/client.go:269]: 这个和 SetProxy的关系是啥

Comment (Duslia) [reply pkg/app/client/client.go]: fix

**Review by welkeyever — APPROVED** [review]

File: pkg/app/client/client.go [hunk 2]
```diff
@@ -576,36 +477,21 @@ func (c *Client) do(ctx context.Context, req *protocol.Request, resp *protocol.R
...
+		hc.SetDynamicConfig(&client.DynamicConfig{
+			Addr:     h,
+			ProxyURI: proxyURI,
+			IsTLS:    isTLS,
+		})
 		m[h] = hc
 		if len(m) == 1 {
 			startCleaner = true
 		}
+
```
Comment (FGYFFFF) [inline pkg/app/client/client.go:494]: 空行要不要去了

File: pkg/app/client/option.go [hunk 3]
```diff
@@ -110,6 +110,34 @@ func WithResponseBodyStream(b bool) config.ClientOption {
...
+}
+
+// WithDisableHeaderNamesNormalizing is used to set whether disable header names normalizing.
+func WithDisableHeaderNamesNormalizing(disable bool) config.ClientOption {
+	return config.ClientOption{F: func(o *config.ClientOptions) {
+		o.DisableHeaderNamesNormalizing = disable
+	}}
+}
+
+// WithName sets client name which used in User-Agent Header
```
Comment (FGYFFFF) [inline pkg/app/client/option.go:124]: 发现这些注释有的带句号有的不太，统一带一下吧

**Review by welkeyever — DISMISSED** [review]

**Review by FGYFFFF — APPROVED** [review]

**Pull request merged**